### PR TITLE
Creates line pivot path planner

### PR DESCRIPTION
**Pull request by sanatd33**
## Description
Describe your pull request.

## Associated / Resolved Issue
Resolves # or [ClickUp card](link-to-clickup-card)

## Design Documents
[Link](link-to-design-doc)

### Test Case 1
1. Step 1
2. Step 2
3. Step 3

**Expected result:**???

## Key Files to Review
Group 1
 * File 1
 * File 2

Group 2
 * File 3
 * File 4

## Review Checklist

- [ ] **Docstrings**: All methods and classes should have the file appropriate docstrings which follow the guidelines in the ["Contributing" page](https://rj-rc-software.readthedocs.io/en/latest/contributing.html) of our docs.
- [ ] **Remove extra print statements**: Any print statements used for debugging should be removed
- [ ] **Tag reviewers**: Tag some people for review and ping them on Slack

## (Optional) Sub-issues (for drafts)
_Note: if you find yourself breaking this PR into many smaller features, it may make sense to break up the PR into logical units based on these features._
- [ ] Step 1
- [ ] Step 2

### CODE REVIEW

File: soccer/src/soccer/planning/planner/line_pivot_path_planner.hpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    // cache the most recent angle change so we know when we're done
+    std::optional<double> cached_angle_change_;
+
+    PathTargetPathPlanner path_target_{};
+
+    Trajectory line(const PlanRequest& request);
+    Trajectory pivot(const PlanRequest& request);
+
+    // TODO(Kevin): ros param this
+    double IS_DONE_ANGLE_CHANGE_THRESH = 1.0;
```
Comment (sid-parikh) [inline soccer/src/soccer/planning/planner/line_pivot_path_planner.hpp]:
```suggestion
    static constexpr double IS_DONE_ANGLE_CHANGE_THRESH {1.0};
```

File: soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    }
+
+    return path;
+}
+
+bool LinePivotPathPlanner::is_done() const {
+    if (!cached_angle_change_.has_value()) {
+        return false;
+    }
+    bool val = abs(cached_angle_change_.value()) < degrees_to_radians(IS_DONE_ANGLE_CHANGE_THRESH);
```
Comment (sid-parikh) [inline soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp]:
```suggestion
    bool val = abs(cached_angle_change_.value()) < degrees_to_radians(static_cast<float>(IS_DONE_ANGLE_CHANGE_THRESH));
```

File: soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    return val;
+}
+
+LinePivotPathPlanner::State LinePivotPathPlanner::next_state(const PlanRequest& plan_request) {
+    const MotionCommand& command = plan_request.motion_command;
+    auto current_point =
+        plan_request.world_state->get_robot(true, plan_request.shell_id).pose.position();
```
Comment (sid-parikh) [inline soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp]:
```suggestion
        plan_request.world_state->get_robot(true, static_cast<int>(plan_request.shell_id)).pose.position();
```

File: soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    auto pivot_point = command.pivot_point;
+
+    double dist_from_point = kRobotRadius;
+    auto target_point = pivot_point + (current_point - pivot_point).normalized(dist_from_point);
+    double vel =
+        plan_request.world_state->get_robot(true, plan_request.shell_id).velocity.linear().mag();
```
Comment (sid-parikh) [inline soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp]:
```suggestion
        plan_request.world_state->get_robot(true, static_cast<int>(plan_request.shell_id)).velocity.linear().mag();
```

File: soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        return LINE;
+    }
+    return current_state_;
+}
+
+Trajectory LinePivotPathPlanner::line(const PlanRequest& plan_request) {
+    const MotionCommand& command = plan_request.motion_command;
+    auto pivot_point = command.pivot_point;
+    auto current_point =
+        plan_request.world_state->get_robot(true, plan_request.shell_id).pose.position();
```
Comment (sid-parikh) [inline soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp]:
```suggestion
        plan_request.world_state->get_robot(true, static_cast<int>(plan_request.shell_id)).pose.position();
```

File: soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    auto final_position = pivot_point + (pivot_point - pivot_target).normalized(radius);
+    std::vector<Point> points;
+
+    // max_speed = max_radians * radius
+    MotionConstraints new_constraints = request.constraints.mot;
+    new_constraints.max_speed =
+        std::min(new_constraints.max_speed, rotation_constraints.max_speed * radius) * .5;
+
+    double start_angle = pivot_point.angle_to(
+        request.world_state->get_robot(true, request.shell_id).pose.position());
```
Comment (sid-parikh) [inline soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp]:
```suggestion
        request.world_state->get_robot(true, static_cast<int>(request.shell_id)).pose.position());
```

File: soccer/src/soccer/planning/planner/pivot_path_planner.cpp [hunk 7]
```diff
@@ -105,7 +105,8 @@ bool PivotPathPlanner::is_done() const {
     if (!cached_angle_change_.has_value()) {
         return false;
     }
-    return cached_angle_change_.value() < IS_DONE_ANGLE_CHANGE_THRESH;
+    bool val = abs(cached_angle_change_.value()) < degrees_to_radians(IS_DONE_ANGLE_CHANGE_THRESH);
```
Comment (sid-parikh) [inline soccer/src/soccer/planning/planner/pivot_path_planner.cpp]:
```suggestion
    bool val = abs(cached_angle_change_.value()) < degrees_to_radians(static_cast<float>(IS_DONE_ANGLE_CHANGE_THRESH));
```

File: soccer/src/soccer/strategy/agent/position/pivot_test.cpp [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            if (check_is_done()) {
+                i++;
+                return IDLE;
+            }
+        }
+        case IDLE: {
+            if (current_play_state_.is_playing()) {
+                if (i % 2 == 0) {
+                    return OUR_GOAL;
+                } else {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/pivot_test.cpp]: note the GH warning here

File: soccer/src/soccer/strategy/agent/position/pivot_test.cpp [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+std::string Pivot::get_current_state() {
+    return std::string{"Pivot"} + std::to_string(static_cast<int>(current_state_));
+}
+
+Pivot::State Pivot::next_state() {
+    // handle transitions between current state
+    switch (current_state_) {
+        case OUR_GOAL: {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/pivot_test.cpp:26]: clang-tidy wants you to collapse these cases. but i'm okay if you don't. it's up to you

File: soccer/src/soccer/strategy/agent/position/pivot_test.hpp [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+#include "position.hpp"
+#include "rj_common/time.hpp"
+#include "rj_constants/constants.hpp"
+#include "rj_geometry/geometry_conversions.hpp"
+#include "rj_geometry/point.hpp"
+#include "seeker.hpp"
+
+namespace strategy {
+
+class Pivot : public Position {
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/pivot_test.hpp:23]: note the clang-tidy warning you need to define some special member functions  see "rule of five" here https://en.cppreference.com/w/cpp/language/rule_of_three

Comment (sid-parikh) [reply soccer/src/soccer/strategy/agent/position/pivot_test.hpp:23]: note that you can explicitly `= default;` them

File: soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp [hunk 11]
```diff
@@ -4,7 +4,7 @@ namespace strategy {
 
 RobotFactoryPosition::RobotFactoryPosition(int r_id) : Position(r_id, "RobotFactoryPosition") {
     if (robot_id_ == 0) {
-        current_position_ = std::make_unique<Goalie>(robot_id_);
```
Comment (sid-parikh) [inline soccer/src/soccer/strategy/agent/position/robot_factory_position.cpp]: uncommit this please

**Review by sid-parikh — CHANGES_REQUESTED** [review]
please look at clang-tidy generated warnings in GitHub

**Review by sid-parikh — DISMISSED** [review]

**Review by rishiso — APPROVED** [review]

File: soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    return path;
+}
+
+bool LinePivotPathPlanner::is_done() const {
+    if (!cached_angle_change_.has_value()) {
+        return false;
+    }
+    bool val = abs(cached_angle_change_.value()) <
+               degrees_to_radians(static_cast<float>(is_done_angle_change_thresh_));
+    return val;
```
Comment (rishiso) [inline soccer/src/soccer/planning/planner/line_pivot_path_planner.cpp]:
```suggestion
    return abs(cached_angle_change_.value()) <
               degrees_to_radians(static_cast<float>(is_done_angle_change_thresh_));
```

**Review by sid-parikh — APPROVED** [review]
we need to work on documenting the path planning logic/math

**Pull request merged**